src/main/java/org/jvnet/hudson/test/InboundAgentRule.java

```diff
@@ -321,4 +337,127 @@ public boolean isAlive(String name) {
+        }
+    }
+
+    private static class GetAgentArguments implements RealJenkinsRule.Step2<AgentArguments> {
+
+        private final String name;
+
+        GetAgentArguments(String name) {
+            this.name = name;
+        }
+
+        @Override
+        public AgentArguments run(@NonNull JenkinsRule r) throws Throwable {
+            return get(r, name);
+        }
+        private static AgentArguments get(JenkinsRule r, String name) throws IOException {
+            Node node = r.jenkins.getNode(name);
+            if (node == null) {
+                throw new AssertionError("no such agent: " + name);
+            }
+            SlaveComputer c = (SlaveComputer) node.toComputer();
+            if (c == null) {
+                throw new AssertionError("agent " + node + " has no executor");
+            }
+            JNLPLauncher launcher = (JNLPLauncher) c.getLauncher();
+            List<String> commandLineArgs = Collections.emptyList();
+            if (!launcher.getWorkDirSettings().isDisabled()) {
+                commandLineArgs = launcher.getWorkDirSettings().toCommandLineArgs(c);
+            }
+            File agentJar = new File(r.jenkins.getRootDir(), "agent.jar");
+            if (!agentJar.isFile()) {
+                FileUtils.copyURLToFile(new Slave.JnlpJar("agent.jar").getURL(), agentJar);
+            }
+            return new AgentArguments(r.getURL() + "computer/" + name + "/slave-agent.jnlp", agentJar, c.getJnlpMac(), r.jenkins.getNodes().size(), commandLineArgs);
+        }
+
+    }
+
+    private static class StopAgent implements RealJenkinsRule.Step {
+        private final String name;
+        StopAgent(String name) {
+            this.name = name;
+        }
+
+        @Override
+        public void run(JenkinsRule r) throws Throwable {
+            stop(r, name);
+        }
+
+        private static void stop(JenkinsRule r, String name) throws InterruptedException {
+            Computer c = r.jenkins.getComputer(name);
+            if (c != null) {
+                while (c.isOnline()) {
+                    Thread.sleep(100);
+                }
+            }
+        }
+    }
+
+    private static class CreateAgent implements RealJenkinsRule.Step2<String> {
+        private final Options options;
+
+        public CreateAgent(Options options) {
+            this.options = options;
+        }
+
+        @Override
+        public String run(JenkinsRule r) throws Throwable {
+            createAgent(r, options);
+            return options.getName();
+        }
+
+        @SuppressFBWarnings(value = {"PATH_TRAVERSAL_IN", "NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE"}, justification = "just for test code")
+        private static Slave createAgent(JenkinsRule r, Options options) throws Descriptor.FormException, IOException, InterruptedException {
+            if (options.getName() == null) {
+                Integer numberOfNodes = r.jenkins.getNodes().size();
```

Comment:
This wrapper type could be a primitive:

```suggestion
                int numberOfNodes = r.jenkins.getNodes().size();
```

Per IntelliJ:

> in some cases, boxing can be source of significant performance penalty, especially in loops.